ompi/runtime/ompi_spc.h

```diff
@@ -0,0 +1,212 @@
+#include "ompi/runtime/params.h"
+#include "opal/mca/timer/timer.h"
+#include "opal/mca/base/mca_base_pvar.h"
+#include "opal/util/argv.h"
+#include "opal/util/show_help.h"
+#include "opal/util/output.h"
+
+#include MCA_timer_IMPLEMENTATION_HEADER
+
+/* INSTRUCTIONS FOR ADDING COUNTERS
+ * 1.) Add a new counter name in the ompi_spc_counters_t enum before
+ *     OMPI_SPC_NUM_COUNTERS below.
+ * 2.) Add corresponding counter name and descriptions to the
+ *     counter_names and counter_descriptions arrays in
+ *     ompi_spc.c  NOTE: The names and descriptions
+ *     MUST be in the same array location as where you added the
+ *     counter name in step 1.
+ * 3.) If this counter is based on a timer, add its enum name to
+ *     the logic for timer-based counters in the ompi_spc_init
+ *     function in ompi_spc.c
+ * 4.) Instrument the Open MPI code base where it makes sense for
+ *     your counter to be modified using the SPC_RECORD macro.
+ *     Note: If your counter is timer-based you should use the
+ *     SPC_TIMER_START and SPC_TIMER_STOP macros to record
+ *     the time in cycles to then be converted to microseconds later
+ *     in the ompi_spc_get_count function when requested by MPI_T
+ */
+
+/* This enumeration serves as event ids for the various events */
+enum ompi_spc_counters_t {
+    OMPI_SEND,
+    OMPI_BSEND,
+    OMPI_RSEND,
+    OMPI_SSEND,
+    OMPI_RECV,
+    OMPI_MRECV,
+    OMPI_ISEND,
+    OMPI_IBSEND,
+    OMPI_IRSEND,
+    OMPI_ISSEND,
+    OMPI_IRECV,
+    OMPI_SENDRECV,
+    OMPI_SENDRECV_REPLACE,
+    OMPI_PUT,
+    OMPI_RPUT,
+    OMPI_GET,
+    OMPI_RGET,
+    OMPI_PROBE,
+    OMPI_IPROBE,
+    OMPI_BCAST,
+    OMPI_IBCAST,
+    OMPI_REDUCE,
+    OMPI_IREDUCE,
+    OMPI_IREDUCE_SCATTER,
+    OMPI_IREDUCE_SCATTER_BLOCK,
+    OMPI_ALLREDUCE,
+    OMPI_IALLREDUCE,
+    OMPI_SCAN,
+    OMPI_ISCAN,
+    OMPI_SCATTER,
+    OMPI_SCATTERV,
+    OMPI_ISCATTER,
+    OMPI_ISCATTERV,
+    OMPI_GATHER,
+    OMPI_GATHERV,
+    OMPI_IGATHER,
+    OMPI_ALLTOALL,
+    OMPI_IALLTOALL,
+    OMPI_IALLTOALLV,
+    OMPI_IALLTOALLW,
+    OMPI_NEIGHBOR_ALLTOALL,
+    OMPI_NEIGHBOR_ALLTOALLV,
+    OMPI_NEIGHBOR_ALLTOALLW,
+    OMPI_ALLGATHER,
+    OMPI_ALLGATHERV,
+    OMPI_IALLGATHER,
+    OMPI_IALLGATHERV,
+    OMPI_NEIGHBOR_ALLGATHER,
+    OMPI_NEIGHBOR_ALLGATHERV,
+    OMPI_TEST,
+    OMPI_TESTALL,
+    OMPI_TESTANY,
+    OMPI_TESTSOME,
+    OMPI_WAIT,
+    OMPI_WAITALL,
+    OMPI_WAITANY,
+    OMPI_WAITSOME,
+    OMPI_BARRIER,
+    OMPI_IBARRIER,
+    OMPI_WTIME,
+    OMPI_CANCEL,
+    OMPI_BYTES_RECEIVED_USER,
+    OMPI_BYTES_RECEIVED_MPI,
+    OMPI_BYTES_SENT_USER,
+    OMPI_BYTES_SENT_MPI,
+    OMPI_BYTES_PUT,
+    OMPI_BYTES_GET,
+    OMPI_UNEXPECTED,
+    OMPI_OUT_OF_SEQUENCE,
+    OMPI_MATCH_TIME,
+    OMPI_UNEXPECTED_IN_QUEUE,
+    OMPI_OOS_IN_QUEUE,
+    OMPI_MAX_UNEXPECTED_IN_QUEUE,
+    OMPI_MAX_OOS_IN_QUEUE,
+    OMPI_SPC_NUM_COUNTERS /* This serves as the number of counters.  It must be last. */
```

Comment:
My prior comment actually intended you to put `SPC_` in *all* of these enums.  `OMPI_SPC_NUM_COUNTERS` was a single example, but it also applies to `OMPI_SPC_MAX_OOS_IN_QUEUE`, `OMPI_SPC_MAX_UNEXPECTED_IN_QUEUE`, ...etc.  I know this is a PITA, but this is a limitation of C -- we need to have decent prefixes to our values as a primitive namespace.  Sorry.  😦